File: artifact/verifiers/python/verifier.py
#!/usr/bin/env python3
from __future__ import annotations

import json
import sys

IMPLEMENTATION = "python-imperative/0.1.0"


def subset(left: list[str], right: list[str]) -> bool:
    return set(left).issubset(set(right))


def evaluate(tx: dict) -> str:
    rp, p, r = tx["rp"], tx["presentation"], tx["reliance"]
    if not p["parse_ok"]:
        return "R_PARSE_INVALID"
    if not rp["registered"]:
        return "R_RP_UNREGISTERED"
    if rp["registration_status"] != "active":
        return "R_RP_SUSPENDED"
    if rp["authenticated_rp_id"] != rp["registered_rp_id"]:
        return "R_RP_IDENTITY_MISMATCH"
    if rp["transaction_purpose"] not in rp["registered_purposes"]:
        return "R_PURPOSE_MISMATCH"
    if not subset(rp["requested_attributes"], rp["registered_attributes"]):
        return "R_ATTRIBUTE_OVER_REQUEST"
    if not p["issuer_signature_valid"]:
        return "R_ISSUER_SIGNATURE_INVALID"
    if not p["all_disclosures_bound"]:
        return "R_DISCLOSURE_UNBOUND"
    if not p["holder_binding_valid"]:
        return "R_HOLDER_BINDING_INVALID"
    if p["audience"] != p["expected_audience"]:
        return "R_AUDIENCE_MISMATCH"
    if p["nonce"] != p["expected_nonce"]:
        return "R_NONCE_MISMATCH"
    age = p["evaluated_at_epoch"] - p["issued_at"]
    if age < 0 or age > p["max_age_seconds"]:
        return "R_PRESENTATION_STALE"
    if p["credential_status"] != "active":
        return "R_CREDENTIAL_INACTIVE"
    if not p["issuer_trusted"]:
        return "R_ISSUER_UNTRUSTED"
    if not subset(r["used_attributes"], r["approved_attributes"]):
        return "R_USE_EXCEEDS_APPROVAL"
    if not r["lawful_basis_recorded"]:
        return "R_LAWFUL_BASIS_UNRECORDED"
    if not r["decision_log_complete"] or not r["verifier_version"] or not r["policy_version"]:
        return "R_AUDIT_EVIDENCE_INCOMPLETE"
    return "A_PROFILE_CONFORMANT"


def main() -> None:
    corpus = json.load(sys.stdin)
    results = []
    for vector in corpus["vectors"]:
        reason = evaluate(vector["transaction"])
        results.append({
            "id": vector["id"],
            "implementation": IMPLEMENTATION,
            "verdict": "ACCEPT" if reason.startswith("A_") else "REJECT",
            "reason": reason,
        })
    json.dump({"implementation": IMPLEMENTATION, "results": results}, sys.stdout, separators=(",", ":"))
    sys.stdout.write("\n")


if __name__ == "__main__":
    main()

File: artifact/verifiers/javascript/verifier.mjs
#!/usr/bin/env node
import fs from "node:fs";

const IMPLEMENTATION = "javascript-functional/0.1.0";
const corpus = JSON.parse(fs.readFileSync(0, "utf8"));
const subset = (left, right) => left.every((item) => right.includes(item));

function evaluate(tx) {
  const { rp, presentation: p, reliance: r } = tx;
  const age = p.evaluated_at_epoch - p.issued_at;
  const rules = [
    [!p.parse_ok, "R_PARSE_INVALID"],
    [!rp.registered, "R_RP_UNREGISTERED"],
    [rp.registration_status !== "active", "R_RP_SUSPENDED"],
    [rp.authenticated_rp_id !== rp.registered_rp_id, "R_RP_IDENTITY_MISMATCH"],
    [!rp.registered_purposes.includes(rp.transaction_purpose), "R_PURPOSE_MISMATCH"],
    [!subset(rp.requested_attributes, rp.registered_attributes), "R_ATTRIBUTE_OVER_REQUEST"],
    [!p.issuer_signature_valid, "R_ISSUER_SIGNATURE_INVALID"],
    [!p.all_disclosures_bound, "R_DISCLOSURE_UNBOUND"],
    [!p.holder_binding_valid, "R_HOLDER_BINDING_INVALID"],
    [p.audience !== p.expected_audience, "R_AUDIENCE_MISMATCH"],
    [p.nonce !== p.expected_nonce, "R_NONCE_MISMATCH"],
    [age < 0 || age > p.max_age_seconds, "R_PRESENTATION_STALE"],
    [p.credential_status !== "active", "R_CREDENTIAL_INACTIVE"],
    [!p.issuer_trusted, "R_ISSUER_UNTRUSTED"],
    [!subset(r.used_attributes, r.approved_attributes), "R_USE_EXCEEDS_APPROVAL"],
    [!r.lawful_basis_recorded, "R_LAWFUL_BASIS_UNRECORDED"],
    [!r.decision_log_complete || !r.verifier_version || !r.policy_version, "R_AUDIT_EVIDENCE_INCOMPLETE"],
  ];
  return rules.find(([failed]) => failed)?.[1] ?? "A_PROFILE_CONFORMANT";
}

const results = corpus.vectors.map((vector) => {
  const reason = evaluate(vector.transaction);
  return {
    id: vector.id,
    implementation: IMPLEMENTATION,
    verdict: reason.startsWith("A_") ? "ACCEPT" : "REJECT",
    reason,
  };
});

process.stdout.write(`${JSON.stringify({ implementation: IMPLEMENTATION, results })}\n`);
